### PR TITLE
Quantity CSV export improvements 

### PR DESCRIPTION
**Pull request by mstenta**
The goal of this PR is to improve the Quantity CSV export functionality in farmOS.

- [x] Add support for Fraction fields in CSV exports.
- [x] Add an "Export CSV" bulk action to the `farm_log_quantity` View (depends on, and includes commits from, #860)
- [x] Remove the existing "Export CSV" link and display from the View.
- [x] Append log fields to Quantity CSV exports: `id`, `status`, `timestamp`, `type`, `name`, `assets`
- [x] Add an "Export Quantity CSV" bulk action to the `farm_log` View, so that it's possible to export quantities directly from Views of logs.

### CODE REVIEW

File: modules/core/log/modules/quantity/src/Normalizer/QuantityNormalizer.php
Comment (paul121) [inline modules/core/log/modules/quantity/src/Normalizer/QuantityNormalizer.php:1]: @mstenta I implemented what you started here. Note how this prepends the log fields to the normalized data array and does a recursive normalize on the `log.timestamp`, using the same `$context` to maintain the human-readable timestamps. Works well!

I wasn't sure about 'bailing' if the quantity didn't reference a log.. returning an empty `$data` array might cause other issues? But really we should be able to expect that quantities will have a log so I don't think this is an issue.

One issue with this approach is that if you had quantities associated with other entity types eg "asset quantities", then it would be hard for modules to provide their own normalizers to handle the quantity -> csv normalization correctly and include "asset" data instead of logs. Not sure if this is something we can solve. Perhaps quantities could have a `serialization` event and allow modules to tack on additional metadata as needed?

File: modules/core/export/modules/csv/src/Plugin/Action/LogQuantityCsv.php [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+
+/**
+ * Action that exports a CSV file of log quantities.
+ *
+ * @Action(
+ *   id = "farm_export_csv:log_quantity",
+ *   action_label = @Translation("Export Quantities CSV"),
+ *   type = "log",
+ *   confirm_form_route_name = "entity.quantity.csv_form",
+ * )
```
Comment (paul121) [inline modules/core/export/modules/csv/src/Plugin/Action/LogQuantityCsv.php:19]: This is a bit of a clever way to implement this Log Quantity CSV export... I found that we could reuse the `entity.quantity.csv_form` entirely, all we need to do is save the `log.quantities` into the tempstore instead of the `log` itself. This means we only need to implement this single Action plugin (not using a deriver) and don't need to duplicate almost all of the logic from `EntityCsvActionForm`.

This could create a little of bit of weird UX though... When the user selects/checks logs and clicks the "Export Quantities CSV" action, it will take them to the same "Export a CSV of [X] quantities?" confirmation form as you would get when using "Export CSV" action from a quantities view. It will not mention anything of the logs they selected. It's a little odd, but really it's just elegant and might not be too confusing to users anyways. If you do click 'Cancel' it takes you back to the correct destination Log page, not Log Quantities.

The major downside I foresee with this approach is that it just might not scale and exporting many logs with many quantities might crash, even if we are able to add a Batch API export to the general `EntityCsvActionForm`. This is simply because loading all the log quantities during the action plugin's `executeMultiple()` step can't be included in a Batch operation. Ideally, the Batch export for Log Quantities would handle a few or even a single _log_ (not quantity) in each batch operation. In this case I think the solution is providing a new Export Action Form that handles this Log + Quantity export more elegantly. I think this could be added/changed in the future without breaking change to the Action config entity added here.

File: modules/core/log/modules/quantity/src/Normalizer/QuantityNormalizer.php
Comment (mstenta) [reply modules/core/log/modules/quantity/src/Normalizer/QuantityNormalizer.php:1]: Looks good @paul121!

> I wasn't sure about 'bailing' if the quantity didn't reference a log.. returning an empty $data array might cause other issues?

I like that it will only modify `$data` if it finds a log associated with it. Otherwise the original `$data` will be returned. That feels like the right behavior, and means that if a log is associated with another entity (or is orphaned), the export will still work.

> Perhaps quantities could have a `serialization` event and allow modules to tack on additional metadata as needed?

Yea this seems like the only way to handle it. That can be a future task.

File: modules/core/export/modules/csv/src/Plugin/Action/LogQuantityCsv.php (hunk 1 above)
Comment (mstenta) [reply modules/core/export/modules/csv/src/Plugin/Action/LogQuantityCsv.php:19]: > This is a bit of a clever way to implement this Log Quantity CSV export...

Brilliant! That works well!

> This could create a little of bit of weird UX though...

I don't think that's too weird. If we find that it causes too much confusion we can consider ways to make that more clear. But I think this is fine for now.

> The major downside I foresee with this approach is that it just might not scale

Hmm this is a good point. I agree we can merge this for now and address that potential issue in the future.

**Comment by mstenta** [comment]
Force pushed to fix two things:

- PHPCS error.
- Fixed config dependency name (`farm_log_quantity` instead of `log_quantity`).

**Comment by mstenta** [comment]
Also added `enforced` dependency to the log quantity action config for consistency (not sure if this is actually necessary for `optional` configs, but that's a separate question IMO).

**Comment by mstenta** [comment]
Realized on Rothamsted dev call today that this implementation misses something that is critical for their use-case. It traces back to issue #805 and PR #842.

Tl;dr: when multiple quantities are exported and they are different type (eg: `standard` and `material`), then all the bundle-specific fields are omitted from the export.

Rothamsted needs the `material_type` to be included in exports.

**Comment by mstenta** [comment]
Figured out the failing tests. See the last commit message for details @paul121 - this is ready for final review.

**Review by paul121 — APPROVED** [review]
Interesting. This seems like a fine workaround!
